File: dp/213.py
#因为首尾相连, 考虑第一位抢或不抢 两种情况分开
class Solution:
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        if len(nums) == 2:
            return max(nums)
        l = []
        l1 = []
        # while nums[1] > nums[0] or nums[len(nums)-1] > nums[0]:
        #     temp = nums[0]
        #     del nums[0]
        #     nums.append(temp)
        # print(nums)
        # l = [0] * len(nums)
        l.append(nums[0])
        l1.append(0)
        l.append(max(nums[0], nums[1]))
        l1.append(nums[1])
        for i in range(2, len(nums)):
            if i == len(nums) - 1:
                l.append(l[i-1])
            else:
                l.append(max(l[i-2] + nums[i], l[i-1]))
            if i == 2:
                l1.append(max(l1[i-1], nums[i]))
            else:
                l1.append(max(l1[i-2] + nums[i], l1[i-1]))

        return max(max(l), max(l1))
                

File: dp/213.cc
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.empty()) {
            return 0;
        }
        if (nums.size() == 1) {
            return nums[0];
        }
        return max(robRange(nums, 0, nums.size()-2), robRange(nums, 1, nums.size()-1));
    }
    
private:
    int robRange(vector<int>& nums, int start, int end) {
        int dp_i_1 = 0;
        int dp_i_2 = 0;
        int dp = 0;
        for (int i = start; i <= end; i++) {
            dp = max(dp_i_1+nums[i], dp_i_2);
            dp_i_1 = dp_i_2;
            dp_i_2 = dp;
        }
        return dp_i_2;
    }
};
